src/datachain/lib/dc.py

```diff
@@ -331,22 +346,80 @@ def from_storage(
             chain = DataChain.from_storage("s3://my-bucket/my-dir")
             ```
         """
-        func = get_file(type)
-        return (
-            cls(
-                path,
+        file_type = get_file_type(type)
+
+        if client_config is None:
+            client_config = {}
+
+        if anon:
+            client_config["anon"] = True
+
+        session = Session.get(
+            session,
+            catalog=kwargs.get("catalog"),
+            client_config=client_config,
+            in_memory=in_memory,
+        )
+
+        client, path = Client.parse_url(
+            uri, session.catalog.cache, **session.catalog.client_config
+        )
+
+        # clean path without globs
+        lst_path = (
+            posixpath.dirname(path)
+            if glob.has_magic(os.path.basename(os.path.normpath(path)))
+            or client.fs.isfile(uri)
+            else path
+        )
+        lst_uri = f"{client.uri}/{lst_path.lstrip('/')}"
+
+        ds_name = listing_dataset_name(client.uri, posixpath.join(lst_path, ""))
+
+        for ds in cls.datasets(
+            session=session, in_memory=in_memory, include_listing=True
+        ).collect("dataset"):
+            if listing_expired(ds.created_at):  # type: ignore[union-attr]
+                continue
+            if (
+                listing_subset(ds.name, ds_name)  # type: ignore[union-attr]
+                and not update
+            ):
+                # we can use found listing as it contains the one from input
+                dc = cls.from_dataset(ds.name, **kwargs)  # type: ignore[union-attr]
```

Comment:
code looks very repetitive with the https://github.com/iterative/datachain/pull/294/files#diff-0b1cfd3d89c057cc57811e07eb20298eeaaee60c54cc66e5128f618d144f42eeR410-R413 - can we please try to use the same?

otherwise it raises questions right away - e.g. why in one case we pass sessions, and don't do this in the second case, etc 

+ plus it'll be harder to main

Reply:
You are right, `session` arg is missing in that line. Adding it.

Reply:
can we please deduplicate and simplify this?

Reply:
Refactored